recommenders/models/ncf/dataset.py

```diff
@@ -1,321 +1,450 @@
+    ):
+        self.filename = filename
+        self.col_user = col_user
+        self.col_item = col_item
+        self.col_rating = col_rating
+        self.col_test_batch = col_test_batch
+        self.expected_fields = [self.col_user, self.col_item, self.col_rating]
+        if self.col_test_batch is not None:
+            self.expected_fields.append(self.col_test_batch)
+        self.binary = binary
+        self.user2id, self.item2id, self.id2user, self.id2item = None, None, None, None
+        self._init_data()
+
+    @property
+    def users(self):
+        return self.user2id.keys()
+
+    @property
+    def items(self):
+        return self.item2id.keys()
+
+    @property
+    def end_of_file(self):
+        return (self.line_num > 0) and self.next_row is None
+
+    def __iter__(self):
+        return self
+
+    def __enter__(self, *args):
+        self.file = open(self.filename, 'r', encoding='UTF8')
+        self.reader = csv.DictReader(self.file)
+        self._check_for_missing_fields(self.expected_fields)
+        self.line_num = 0
+        self.row, self.next_row = None, None
+        return self
+
+    def __exit__(self, *args):
+        self.file.close()
+        self.reader = None
+        self.line_num = 0
+        self.row, self.next_row = None, None
+
+    def __next__(self):
+        if self.next_row:
+            self.row = self.next_row
+        elif self.line_num == 0:
+            self.row = self._extract_row_data(next(self.reader, None))
+            if self.row is None:
+                raise Exception("{} is empty.".format(self.filename))
+        else:
+            raise StopIteration  # end of file
+
+        self.next_row = self._extract_row_data(next(self.reader, None))
+        self.line_num += 1
+
+        return self.row
+
+    def _check_for_missing_fields(self, fields_to_check):
+        missing_fields = set(fields_to_check).difference(set(self.reader.fieldnames))
+        if len(missing_fields):
+            raise ValueError("Columns {} not in header of file {}".format(missing_fields, self.filename))
+
+    def _extract_row_data(self, row):
+        if row is None:
+            return row
+        user = int(row[self.col_user])
+        item = int(row[self.col_item])
+        rating = float(row[self.col_rating])
+        if self.binary:
+            rating = float(rating > 0)
+        test_batch = None
+        if self.col_test_batch:
+            test_batch = int(row[self.col_test_batch])
+        return {self.col_user: user, self.col_item: item, self.col_rating: rating, self.col_test_batch: test_batch}
+
+    def _init_data(self):
+        # Compile lists of unique users and items, assign IDs to users and items,
+        # and ensure file is sorted by user (and batch index if test set)
+        logger.info("Indexing {} ...".format(self.filename))
+        with self:
+            user_items = []
+            self.item2id, self.user2id = OrderedDict(), OrderedDict()
```

Comment:
Do you need an `OrderedDict`? I don't see any of its specialized functionality used in the code.

Reply:
I'm using OrderedDict to be memory efficient. It serves two purposes:
1. I need a dict to hold the mapping between original user/item values and their IDs
2. I need an ordered list to hold users so I can iterate through them in the order they appear in the training data file
Ordered dict allows me to achieve both these requirements in one object (the `users` attribute can be obtained from the `users` property which just retrieves the OrderedDict keys). If I don't use OrderedDict I think I would have to have both a dictionary and a list each containing all users.

Reply:
`users` returns the output of `keys()`. You could do the same with the standard `dict`, couldn't you?
The LIFO or FIFO functionality provided by OrderedDict isn't used as far as I can see.

Reply:
Yes, we require 3.6 - 3.8 in fact 
https://github.com/microsoft/recommenders/blob/9898ec6f5bc611a23657a6f3395ef9403f210344/setup.py#L142

Reply:
Actually, maybe it's better to keep OrderedDict just to be on the safe side. Insertion order became a platform independent feature in Python 3.7 https://mail.python.org/pipermail/python-dev/2017-December/151283.html. So maybe there will be a problem if someone uses 3.6 elsewhere than CPython.